File: src/lib/recurrence.ts
import type { RecurringTransaction } from '../types'

const pad = (n: number) => String(n).padStart(2, '0')

/**
 * Returns all YYYY-MM-DD date strings when a recurring rule fires
 * within the given calendar month (1-indexed month).
 */
export function expandRecurring(
  rule: RecurringTransaction,
  year: number,
  month: number
): string[] {
  const daysInMonth = new Date(year, month, 0).getDate()
  const dateStr = (d: number) => `${year}-${pad(month)}-${pad(d)}`

  switch (rule.recurrence_type) {
    case 'monthly_fixed': {
      if (rule.day_of_month == null) return []
      const day = Math.min(rule.day_of_month, daysInMonth)
      return [dateStr(day)]
    }

    case 'weekly': {
      if (rule.day_of_week == null) return []
      const firstDow = new Date(year, month - 1, 1).getDay()
      const startDay = ((rule.day_of_week - firstDow + 7) % 7) + 1
      const result: string[] = []
      for (let d = startDay; d <= daysInMonth; d += 7) {
        result.push(dateStr(d))
      }
      return result
    }

    case 'biweekly': {
      if (rule.day_of_week == null || rule.biweekly_anchor == null) return []
      // anchor must be a valid YYYY-MM-DD on the correct day of week
      const anchor = new Date(rule.biweekly_anchor + 'T00:00:00')
      const firstDow = new Date(year, month - 1, 1).getDay()
      const startDay = ((rule.day_of_week - firstDow + 7) % 7) + 1
      const result: string[] = []
      for (let d = startDay; d <= daysInMonth; d += 7) {
        const candidate = new Date(year, month - 1, d)
        const diffDays = Math.round(
          (candidate.getTime() - anchor.getTime()) / 86_400_000
        )
        if (diffDays % 14 === 0) result.push(dateStr(d))
      }
      return result
    }

    case 'yearly': {
      if (rule.month == null || rule.day_of_month == null) return []
      if (rule.month !== month) return []
      const day = Math.min(rule.day_of_month, daysInMonth)
      return [dateStr(day)]
    }

    case 'monthly_nth_weekday': {
      if (rule.day_of_week == null || rule.nth_week == null) return []
      const firstDow = new Date(year, month - 1, 1).getDay()
      const startDay = ((rule.day_of_week - firstDow + 7) % 7) + 1
      const occurrences: number[] = []
      for (let d = startDay; d <= daysInMonth; d += 7) {
        occurrences.push(d)
      }
      if (rule.nth_week === -1) {
        const last = occurrences.at(-1)
        return last != null ? [dateStr(last)] : []
      }
      const idx = rule.nth_week - 1
      return occurrences[idx] != null ? [dateStr(occurrences[idx])] : []
    }
  }
}

/**
 * Returns the 1-based occurrence number of `date` for this rule, counting
 * every firing from `rule.start_date` (inclusive) through `date` (inclusive).
 * Requires `rule.start_date` to be set — without an anchor there's no
 * well-defined "1st occurrence" to count from. Returns 0 if `date` isn't
 * on/after the start date.
 */
export function countOccurrencesThrough(
  rule: RecurringTransaction,
  date: string
): number {
  if (!rule.start_date || date < rule.start_date) return 0
  const [startY, startM] = rule.start_date.split('-').map(Number)
  const [toY, toM] = date.split('-').map(Number)

  let count = 0
  let y = startY, m = startM
  while (y < toY || (y === toY && m <= toM)) {
    for (const d of expandRecurring(rule, y, m)) {
      if (d >= rule.start_date && d <= date) count++
    }
    m++
    if (m > 12) { m = 1; y++ }
  }
  return count
}

File: ios/BudgetBuddy/Logic/Recurrence.swift
import Foundation

enum Recurrence {
    /// Returns all `YYYY-MM-DD` strings within `(year, month)` (1-indexed) when
    /// the rule fires. Mirrors src/lib/recurrence.ts line for line so the iOS
    /// calendar matches the web app's calendar exactly.
    static func expand(_ rule: RecurringTransaction, year: Int, month: Int) -> [String] {
        let cal = Calendar.gregorian
        let daysInMonth = cal.daysInMonth(year: year, month: month)

        func dateStr(_ d: Int) -> String {
            String(format: "%04d-%02d-%02d", year, month, d)
        }

        switch rule.recurrence_type {
        case .monthly_fixed:
            guard let dom = rule.day_of_month else { return [] }
            return [dateStr(min(dom, daysInMonth))]

        case .weekly:
            guard let dow = rule.day_of_week else { return [] }
            let firstDow = cal.dayOfWeek(year: year, month: month, day: 1)
            let startDay = ((dow - firstDow + 7) % 7) + 1
            return stride(from: startDay, through: daysInMonth, by: 7).map(dateStr)

        case .biweekly:
            guard let dow = rule.day_of_week,
                  let anchorStr = rule.biweekly_anchor,
                  let anchor = Calendar.parseYMD(anchorStr) else { return [] }
            let firstDow = cal.dayOfWeek(year: year, month: month, day: 1)
            let startDay = ((dow - firstDow + 7) % 7) + 1
            var result: [String] = []
            var d = startDay
            while d <= daysInMonth {
                if let candidate = cal.date(year: year, month: month, day: d) {
                    let diff = Int((candidate.timeIntervalSince(anchor) / 86_400).rounded())
                    if diff % 14 == 0 { result.append(dateStr(d)) }
                }
                d += 7
            }
            return result

        case .yearly:
            guard let m = rule.month, let dom = rule.day_of_month, m == month else { return [] }
            return [dateStr(min(dom, daysInMonth))]

        case .monthly_nth_weekday:
            guard let dow = rule.day_of_week, let nth = rule.nth_week else { return [] }
            let firstDow = cal.dayOfWeek(year: year, month: month, day: 1)
            let startDay = ((dow - firstDow + 7) % 7) + 1
            let occurrences = Array(stride(from: startDay, through: daysInMonth, by: 7))
            if nth == -1 {
                return occurrences.last.map { [dateStr($0)] } ?? []
            }
            let idx = nth - 1
            guard idx >= 0, idx < occurrences.count else { return [] }
            return [dateStr(occurrences[idx])]
        }
    }
}

extension Calendar {
    /// Gregorian calendar pinned to the user's current timezone — matches
    /// JavaScript `new Date(y, m-1, d)` behavior used in the web app.
    static var gregorian: Calendar {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = .current
        return c
    }

    func daysInMonth(year: Int, month: Int) -> Int {
        guard let d = date(year: year, month: month, day: 1),
              let r = range(of: .day, in: .month, for: d) else { return 30 }
        return r.count
    }

    /// Sunday=0 .. Saturday=6 (matching JavaScript `Date.getDay()`).
    func dayOfWeek(year: Int, month: Int, day: Int) -> Int {
        guard let d = date(year: year, month: month, day: day) else { return 0 }
        return component(.weekday, from: d) - 1
    }

    func date(year: Int, month: Int, day: Int) -> Date? {
        var c = DateComponents()
        c.year = year; c.month = month; c.day = day
        return self.date(from: c)
    }

    /// Parse `YYYY-MM-DD` at local midnight (matches `new Date('YYYY-MM-DDT00:00:00')`).
    static func parseYMD(_ s: String) -> Date? {
        let parts = s.split(separator: "-")
        guard parts.count == 3,
              let y = Int(parts[0]), let m = Int(parts[1]), let d = Int(parts[2])
        else { return nil }
        return Calendar.gregorian.date(year: y, month: m, day: d)
    }

    static func ymdString(_ d: Date) -> String {
        let c = Calendar.gregorian.dateComponents([.year, .month, .day], from: d)
        return String(format: "%04d-%02d-%02d", c.year ?? 0, c.month ?? 0, c.day ?? 0)
    }

    static func todayYMD() -> String { ymdString(Date()) }
}
